firebase-inappmessaging/src/main/java/com/google/firebase/inappmessaging/internal/ApiClient.java

```diff
@@ -89,28 +89,35 @@ static FetchEligibleCampaignsResponse createCacheExpiringResponse() {
     }
     Logging.logi(FETCHING_CAMPAIGN_MESSAGE);
     providerInstaller.install();
-    return firebaseInstanceId
-        .getInstanceId()
-        .continueWith(
-            instanceIdResultTask -> {
-              InstanceIdResult instanceIdResult = instanceIdResultTask.getResult();
-              if (instanceIdResult == null) {
-                Logging.logw("InstanceID is null, not calling backend");
-                return createCacheExpiringResponse();
-              }
-              return withCacheExpirationSafeguards(
-                  grpcClient
-                      .get()
-                      .fetchEligibleCampaigns(
-                          FetchEligibleCampaignsRequest.newBuilder()
-                              // The project Id we expect is the gcm sender id
-                              .setProjectNumber(firebaseApp.getOptions().getGcmSenderId())
-                              .addAllAlreadySeenCampaigns(
-                                  impressionList.getAlreadySeenCampaignsList())
-                              .setClientSignals(getClientSignals())
-                              .setRequestingClientApp(getClientAppInfo(instanceIdResult))
-                              .build()));
-            });
+    return firebaseInstallations
+        .getId()
+        .continueWithTask(
+            idTask ->
+                firebaseInstallations
+                    .getToken(false)
```

Comment:
It looks like this method can happen in parallel with the above `getId()` any reason you are doing them sequentially?, i.e.  could you try doing `Tasks.whenAll(idTask, tokenTask).continueWith()`?

Reply:
Nice! done